test/forms/UkimsNumberFormProviderSpec.scala

```diff
@@ -0,0 +1,102 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package forms
+
+import forms.behaviours.StringFieldBehaviours
+import org.scalacheck.Gen
+import play.api.data.{Form, FormError}
+
+class UkimsNumberFormProviderSpec extends StringFieldBehaviours {
+
+  private val formProvider       = new UkimsNumberFormProvider()
+  private val form: Form[String] = formProvider()
+  private val requiredKey        = "ukimsNumber.error.required"
+  private val invalidFormatKey   = "ukimsNumber.error.invalidFormat"
+  private val fieldName          = "ukimsNumber"
+
+  ".ukimsNumber" - {
+
+    "valid UKIMS number" - {
+
+      val ukimsNumberGenerator: Gen[String] = {
+        val prefixGen      = Gen.oneOf("GB", "XI")
+        val firstDigitsGen = Gen.listOfN(12, Gen.numChar).map(_.mkString)
+        val lastDigitsGen  = Gen.listOfN(14, Gen.numChar).map(_.mkString)
+
+        for {
+          prefix      <- prefixGen
+          firstDigits <- firstDigitsGen
+          lastDigits  <- lastDigitsGen
+        } yield s"$prefix$firstDigits$lastDigits"
+      }
+
+      behave like mandatoryField(form, fieldName, requiredError = FormError(fieldName, requiredKey))
+
+      behave like fieldThatBindsValidData(form, fieldName, ukimsNumberGenerator)
+
+    }
+
+    "invalid UKIMS number" - {
+
+      "invalid UKIMS number with invalid length" - {
+
+        val invalidUkimsNumberGeneratorWithInvalidLength: Gen[String] = {
+
+          val prefixGen      = Gen.oneOf("GB", "XI")
+          val firstDigitsGen = Gen.listOfN(12, Gen.numChar).map(_.mkString)
+          val lastDigitsGen  = Gen.listOfN(15, Gen.numChar).map(_.mkString)
+
+          for {
+            prefix      <- prefixGen
+            firstDigits <- firstDigitsGen
+            lastDigits  <- lastDigitsGen
+          } yield s"$prefix$firstDigits$lastDigits"
+        }
+
+        behave like fieldThatErrorsOnInvalidData(
+          form,
+          fieldName,
+          invalidUkimsNumberGeneratorWithInvalidLength,
+          FormError(fieldName, invalidFormatKey)
+        )
+      }
+
+      "invalid UKIMS number with invalid prefix" - {
+
+        val invalidUkimsNumberGeneratorWithInvalidPrefix: Gen[String] = {
+          val prefixGen      = Gen.oneOf("AA", "BB")
+          val firstDigitsGen = Gen.listOfN(12, Gen.numChar).map(_.mkString)
+          val lastDigitsGen  = Gen.listOfN(14, Gen.numChar).map(_.mkString)
+
+          for {
+            prefix      <- prefixGen
+            firstDigits <- firstDigitsGen
+            lastDigits  <- lastDigitsGen
+          } yield s"$prefix$firstDigits$lastDigits"
+        }
+
+        behave like fieldThatErrorsOnInvalidData(
+          form,
+          fieldName,
+          invalidUkimsNumberGeneratorWithInvalidPrefix,
+          FormError(fieldName, invalidFormatKey)
+        )
+
+      }
+    }
+  }
+}
```

Comment:
could we have some more tests?

like invalid first digits, what happens if it is blank, what happens with spaces or dashes

happy to be argued that this is not necessary

Reply:
This make sense

Reply:
New unit test added